### PR TITLE
Carry over segmentation data between frames

### LINKED ISSUE #1815: Artefacts when using low quantizer
While running some encodes at CQ 30, speed 5, Psy tune using rav1e commit 46a64c4 I noticed heavy blocking on one clip and general metrics degradation, with PSNR-HVS-M being particularly affected.

For example:

Closer inspection revealed a variety of more or less subtle artefacts, such as intensity shifting in the Y channel, in most clips encoded using the configuration from above.

Reduced clip for testing: https://mega.nz/#!FloAHajK!_OHewBCUU1D7AJvGYG3mxtWzFZNne17aC5j5bcJnRx0
Exact cmdline: `rav1e -o dist.rav1e.cq30.sp5.ivf --quantizer 30 -s 5 --tune Psychovisual rav1e.sp5.cq30.psy.bugRepro.y4m`

I have been unable to reproduce the issue using any other configuration so far, such as lower or higher CQ, using speed 6, or tune Psnr, at least using the clip above for testing.

Attached are two encodes, a faulty one at CQ 30 and a good one at CQ 31.
[dist.encodes.zip](https://github.com/xiph/rav1e/files/3795677/dist.encodes.zip)

### PR DESCRIPTION
**Pull request by barrbrain**
With the data carried over, we can remove a prior workaround and fix the outstanding desyncs.

Fixes #1815.

### CODE REVIEW

File: src/encoder.rs [hunk 1]
```diff
@@ -3388,6 +3389,19 @@ pub fn encode_show_existing_frame<T: Pixel>(
...
+  let segmentation = if fi.primary_ref_frame == PRIMARY_REF_NONE {
+    None
+  } else {
+    let ref_frame_idx = fi.ref_frames[fi.primary_ref_frame as usize] as usize;
+    let ref_frame = fi.rec_buffer.frames[ref_frame_idx].as_ref();
+    ref_frame.map(|rec| rec.segmentation)
+  };
+
+  // return the retrieved instance if any, a new one otherwise
+  segmentation.unwrap_or_else(|| Default::default())
```
Comment (rzumer) [inline src/encoder.rs]:
```suggestion
  segmentation.unwrap_or_default()
```

**Comment by rzumer** [comment]
Also closes #1738.

File: src/encoder.rs [hunk 2]
```diff
@@ -3388,6 +3389,19 @@ pub fn encode_show_existing_frame<T: Pixel>(
   packet
 }
 
+fn get_initial_segmentation<T: Pixel>(fi: &FrameInvariants<T>) -> SegmentationState {
```
Comment (barrbrain) [inline src/encoder.rs]:
```suggestion
fn get_initial_segmentation<T: Pixel>(
  fi: &FrameInvariants<T>,
) -> SegmentationState {
```

**Comment by barrbrain** [comment]
AWCY results on [scenecuts-720p at low QP](https://beta.arewecompressedyet.com/?job=master%402019-11-01T23%3A43%3A18.808Z&job=segmentation-reference%402019-11-04T13%3A50%3A07.911Z):

|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -0.1821 |     N/A |     N/A |   0.6147 | -1.5040 | -0.5962 |        N/A |

Note that the effect of desync is excluded frmo AWCY metrics at the moment due to using rav1e's reconstruction. Yet removing the workaround is helpful beyond fixing the desync.

**Comment by SmilingWolf** [comment]
I can confirm this fixes the issue on the fragment extracted from one of the affected clips.

Tomorrow I'll run the full set of encodes on all my clips and report what I find.

**Pull request merged**